src/parsers.rs

```diff
@@ -1,20 +1,37 @@
+use std::convert::TryFrom;
 use std::iter::once;
 use std::marker::PhantomData;
 
 use indextree::{Arena, NodeId};
 use jetscii::{bytes, BytesConst};
 use memchr::{memchr, memchr_iter};
-use nom::bytes::complete::take_while1;
+use nom::{bytes::complete::take_while1, IResult};
 
 use crate::config::ParseConfig;
 use crate::elements::{
     block::RawBlock, emphasis::Emphasis, keyword::RawKeyword, radio_target::parse_radio_target,
-    Clock, Comment, Cookie, Drawer, DynBlock, Element, FixedWidth, FnDef, FnRef, InlineCall,
-    InlineSrc, Link, List, ListItem, Macros, Rule, Snippet, Table, TableCell, TableRow, Target,
-    Timestamp, Title,
+    timestamp::parse_timestamp, Clock, Comment, Cookie, Drawer, DynBlock, Element, FixedWidth,
+    FnDef, FnRef, InlineCall, InlineSrc, Link, List, ListItem, Macros, Rule, Snippet, Table,
+    TableCell, TableRow, Target, Title,
 };
 use crate::parse::combinators::lines_while;
 
+// Trait for elements which can be parsed from a string, but retain a lifetime
+// dependence on the string, precluding the use of FromStr or TryFrom.
+pub trait Parse<'a> {
```

Comment:
Sorry for missing this in the first review. What's the purpose of this `Parse` trait. I can see it's only required for `helper_for_parse_element` function. But nom have already provided a combinator named `all_consuming` which makes sure the input is totally consumed.

Reply:
I will replace `helper_for_parse_element` with `all_consuming`, but that's more or less unrelated to the trait.

It is helpful for users of the library to be able to parse things from a string that are not part of an Org document. For example, Timestamp::parse("<2020-01-01>"). I want to provide a simpler API than one that returns an IResult and remainder to make it as similar to TryFrom as possible (I can't use TryFrom because of the lifetime).

Having a trait standardizes this, so that you don't need to figure out you need to call `parse_timestamp` or whatever, or just parse a simple document and then extract your element. I don't think that all elements necessarily need this (it's hard to imagine a user wanting to parse an ad-hoc table), but timestamps and dates (and "simple" elements in general) seem like a convenient feature to provide it for.

So essentially, the goal is that you can call ::try_from on elements without a lifetime, and ::parse on elements with or without one.

We can take it out if you like, and possibly reconsider for a future PR that provides it for more elements.

Reply:
I like the idea of using `TryFrom` trait here. For example, we can impl `TryFrom<'a str>` for `Foo<'a>` and `TryFrom<String>` for `Foo<'static>` which means it can handle both `&str` and `String`. But I'm not sure if we should just return an `IResult` or a `Option` since we have already specified the error value of `IResult` to be a `()`.

Reply:
Interesting, I had tried TryFrom before and could not get the lifetime to work out, but when I write it this way it works. Anyway, if we can use TryFrom that is definitely preferable to adding a trait. I removed Parse and instead implemented TryFrom for both. I'm not sure how to move the string, since the elements may just need to refer to subparts, but we can at least call into_owned.

Do we also need it for both String and &str for the types without a lifetime? Or can I do something like AsRef/Borrow<&str> without running into multiple conflicting implementations?

I think TryFrom has to return a Result. You need to provide the associated type Error. It can just be ().

I have changed the code as discussed here.